rts/lra/lra-coordinator/src/main/java/io/narayana/lra/coordinator/internal/LRARecoveryModule.java

```diff
@@ -122,6 +122,12 @@ private void doRecoverTransaction(Uid recoverUid) {
 
                 if (!inFlight && lra.isRecovering()) {
                     lra.replayPhase2();
+
+                    if (!lra.isRecovering()) {
```

Comment:
would not be better to check for `isFinished` or something similar? This is a bit hard to understand, plus it makes possibility to leak a race condition when `lra.replayPhase2()` does some strange thing like changing the state to `Active` or to some currently unknown state

Reply:
This code is following a well known programming idiom:

```
check condition
run some code that can alter the condition
recheck the condition
```

A race is not possible.

The isRecovering() method tests that the LRA has entered the termination phase and has not yet finished and since this is called from periodic recovery the method name is intuitive - it says "is the LRA still in need of recovery?".

Reply:
The code could follow the well know idiom. My question was rather if it's appropriate for this place. How well understandable is such idiom for the reader.

The `replayPhase2` strives to finish the LRA. From that it could be better fitting to check if the LRA is really finished.

The race is not possible now but what happen when the state model is enhanced and the LRA starts to use new action status. Could be then the `toLRAStatus` falls through to be Active?